### PR TITLE
Refactored pipelines

### PR DESCRIPTION
**Pull request by AhmedBM**
## Description

* Refactored pipelines into stages (see /devops/pipelines/stages)
  *  binary_build.yaml - Produces binaries + packages, performs tests and code coverage
  *  source_build.yaml - Produces SOURCE packages (tar.gz) and RPM spec file and signatures.json
* Removed matrix-strategy builds in-favor of steps - provides the ability to re-run, better analytics/statistics
* Package build pipeline (binary_package.yaml) - refactored to use stages template
* Continuous integration pipeline (source_package.yaml) - refactored to use stages template
  *  Added Ubuntu 18.04 ARM/ARM64 variants to CI validation
* Added MSFT copyright to pipelines and added descriptions to the pipelines

* ## Checklist

- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [X] I added unit-tests to validate my changes. All unit tests are passing.
- [X] I have merged the latest `main` branch prior to this PR submission.
- [X] I submitted this PR against the `main` branch.

### CODE REVIEW

File: devops/pipeline/cleanup_e2etest_infra.yaml [hunk 1]
```diff
@@ -1,8 +1,16 @@
-# Test Environments tagged with environment: osconfig-e2etest should be removed daily
-# The 'osconfig-e2etest-debug' environment tag is used to persist the test infrastructure
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# 
+# End-to-end test environment cleanup pipeline
```
Comment (MariusNi) [inline devops/pipeline/cleanup_e2etest_infra.yaml:4]: >End-to-end test environment cleanup pipeline [](http://example.com/codeflow?start=2&length=44)

E2E tests should not be executed by the binary package pipeline but by the PR (aka: continuous integration) pipeline.

Can we accommodate this much needed change as part of this pipeline refactoring?

File: devops/pipeline/stages/source_build.yaml [hunk 2]
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+#
```
Comment (MariusNi) [inline devops/pipeline/stages/source_build.yaml:3]: Here and elsewhere.

The MIT license is stated at root of the project.
The notices at top of each source and infra file must be the  lone copyright notice.

Please remove the license line from this file and the rest.

File: devops/pipeline/cleanup_e2etest_infra.yaml (hunk 1 above)
Comment (robertschaedler3) [reply devops/pipeline/cleanup_e2etest_infra.yaml:4]: I agree, most important change is the E2E tests.

Comment (AhmedBM) [reply devops/pipeline/cleanup_e2etest_infra.yaml:4]: sure, ill put this as part of another PR

File: devops/pipeline/source_packaging.yaml [hunk 3]
```diff
@@ -1,109 +1,42 @@
...
+    publishArtifact: false
+
+- template: stages/binary_build.yaml
+  parameters:
+    variantName: ubuntu18.04
+    friendlyVariantName: ubuntu1804
+    architecture: arm64
+    containerPath: $(CONTAINER_REGISTRY)/ubuntu18.04-dev-arm64:latest
+    platform: linux/arm64/v8
+    publishArtifact: false
```
[the remark this answers is not in the document]
Comment (AhmedBM) [reply devops/pipeline/source_packaging.yaml:42]: We can, just trying to maximize coverage while minimizing jobs, do you think i should have them all?

File: devops/pipeline/binary_package.yaml [hunk 4]
```diff
@@ -16,231 +19,91 @@ trigger:
...
-    - task: PublishCodeCoverageResults@1
-      displayName: Publish code coverage results
-      condition: eq(variables.coverage, true)
-      inputs:
-        codeCoverageTool: 'Cobertura'
-        summaryFileLocation: $(Build.SourcesDirectory)/build-$(arch)/coverage.xml
-        pathToSources: $(Build.SourcesDirectory)/src
-        #failIfCoverageEmpty: false # Optional
\ No newline at end of file
+- template: stages/source_build.yaml
```
[the remark this answers is not in the document]
Comment (AhmedBM) [reply devops/pipeline/binary_package.yaml:22]: yes, they can be reused in any pipeline thats the reason for the refactor. How you structure your pipeline (use different strategies parallel/matrix) is up-to-you

File: devops/pipeline/source_packaging.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (AhmedBM) [reply devops/pipeline/source_packaging.yaml]: They are all done in parallel, so the cost is just an additional "job". We make use of agent pools that have quite abit of resource (for now) but if we move to our own instance, we might want to limit these.

[the remark this answers is not in the document]
Comment (AhmedBM) [reply devops/pipeline/source_packaging.yaml]: Added all variants except for Ubuntu-20.04/ARM since there seems to be issues with that environment

**Review by MariusNi — APPROVED** [review]
:shipit:

**Review by robertschaedler3 — CHANGES_REQUESTED** [review]
Can we run all the pipelines for this PR before merging?

**Pull request merged**